### DISCUSSION TITLE
Redacting values in CypressError logs

### DISCUSSION THREAD

**Discussion by 403-html** [cypress-io/cypress#25661] [Questions and Help] [answered]
Hi. I'm currently facing an issue with redacting values in `CypressError` logs. I have a solution in place, but I'm not happy with the format and it looks quite messy (no indents, or just a lot of new lines). I was wondering if someone could help me improve the code and make it look more presentable, like native CypressError logs does.

This is the code I have so far

```ts
Cypress.on('fail', (error) => {
  const redactedString = '[redacted]';
  const propertiesToRedact = ['password'];

  error.parsedStack.forEach((frame) => {
    if (frame.message) {
      propertiesToRedact.forEach((property) => {
        frame.message = frame.message.replace(
          new RegExp(`\\"${property}\\":\\".*?\\"`, 'g'),
          `"${property}":"${redactedString}"`,
        );
      });
    }
  });

  const newError = new Error(error.parsedStack.map((frame) => frame.message).join('\n'));
  throw newError;
});
```

**Comment by portyu9** [comment]
Hi @403-html 

I would avoid rebuilding the error from `error.parsedStack`.

`parsedStack` is Cypress-internal data, and creating a new `Error` from it throws away metadata that Cypress uses to render the original error, stack trace, and code frame. That is why the replacement looks much worse than the native `CypressError`.

If you must redact at the `fail` event, modify the original error and then rethrow **that same error**:

```js
Cypress.on('fail', (error) => {
  const redact = (text) => {
    if (!text) return text

    return text.replace(
      /("password"\s*:\s*")[^"]*"/g,
      '$1[redacted]"',
    )
  }

  error.message = redact(error.message)
  error.stack = redact(error.stack)

  throw error
})
```

That preserves far more of Cypress's native error handling than constructing a new `Error`.

However, for secrets, the better approach is to stop the value from reaching Cypress's logs in the first place. For example:

```js
cy.get('#password').type(password, { log: false })
```

or wrap `.type()` with a custom command that logs `*****` instead of the actual value.

That distinction is important because redacting the final failure error does **not** retroactively remove a secret that was already written to a command log, screenshot, video, request log, assertion, or browser console.

So I would use the `fail` handler only as a final safety net and do the primary masking at the command that handles the sensitive value.

Please mark this answer as accepted if it helped, thank you.

**Reply by 403-html** [reply]
Thanks for the answer! A bit late for us though 😄

In the ~3.5 years since I asked this, we went through the whole Cypress third-party blocking situation and eventually abandoned Cypress altogether. So unfortunately I can't really verify this against the original problem anymore (seems valid from architectural standpoint)..

Still, leaving it here might help someone who finds this discussion later.
